### ISSUE TITLE
train on exist model

### ISSUE THREAD

**Issue by josef821** [mittagessen/kraken#232]
i train two language Arabic and English in one model { i create Ground-truth *.png and *.txt with both language } . it works good.
now i want to train more new font to Arabic. i use this command : 
ketos train -i ara_eng.mlmodel train/*.png
when train finished, English not working anymore. i try --resize add , --resize both but its not working.
what should i do for train only new fonts to Arabic to exist train model without losing anything?

**Comment by mittagessen** [comment]
* josef821 :: 2021-01-10 13:31 Sun:
> i train two language Arabic and English in one model { i create Ground-truth *.png and *.txt with both language } . it works good.
> now i want to train more new font to Arabic. i use this command : 
> ketos train -i ara_eng.mlmodel train/*.png
> when train finished, English not working anymore. i try --resize add , --resize both but its not working.
> what should i do for train only new fonts to Arabic to exist train model without losing anything?

You can't. The process is fuzzy as the final layer has a softmax
activation which normalizes the activations so that they sum to 1. If
entries of that layer get removed or added it throws all the other
weights a bit off, as the proportion between the weights in the
normalization term changes. Usually this is fairly non-destructive if
only a few characters get added or removed but larger changes cause
enoug degradation that already learned behavior is erased.

**Comment by mittagessen** [comment]
Arrgh sorry, I didn't read correctly. As you're first training
english+arabic and then try to adapt to a new dataset with arabic only
the net just forgets everything it learned on Latin writing (with
--resize add) or it gets deleted explicitly in the preprocessing (with
--resize both). The transfer learning in kraken more or less only allows
you to retarget the learned distribution to a new one with less training
data than learning from scratch. There's no way to protect already
learned information. An ensemble method might be of use for that but we
haven't implemented one, yet.

**Comment by josef821** [comment]
thank you for quick reply.
can join several models into one?

**Comment by mittagessen** [comment]
* josef821 :: 2021-01-11 14:36 Mon:
> thank you for quick reply.  can join several models into one?

Through the new segmenter, yes. The process to have multi-script
capability is to have the segmenter split the page into lines for script
A and script B and then run separate models on the different scripts (I
think I already wrote you how to do that from the command line).
Otherwise you could always train everything together but as you've
experienced that makes 'partial' adaptation impossible as the net
unlearns the other domain.

That's a general problem with neural networks. I believe tesseract 4
allows model combination but runs all models and then picks the
'correct' output for each word based on the model dictionaries which is
rather crude to say it kindly. I doubt it works particularly well,
especially when mixing a low scriptal combinatorial coverage language
(English) with a high scriptal coverage one such as Arabic (or any
written in an abjad I'd assume).

**Comment by josef821** [comment]
new segmenter will detect every line language, that's good but sometimes line will like this :
ثاني أغنية فردية له “ Ben “ التي تحكي
some English text will come in Arabic text.
i need to train both of them for fix that.
thank you for your help.

**Comment by mittagessen** [comment]
* josef821 :: 2021-01-12 08:55 Tue:
> new segmenter will detect every line language, that's good but sometimes line will like this :
> ثاني أغنية فردية له “ Ben “ التي تحكي
> some English text will come in Arabic text.
> i need to train both of them for fix that.
> thank you for your help.

Ah yeah. For mixed lines we had an intra-line script detection in the
past but the new neural backend broke it unfixably. Joint training is
really the best in that case.
